### PR TITLE
Run plank controller sync with bounded concurrency

### PR DESCRIPTION
**Pull request by stevekuznetsov**
When there are issues with the underlying Jenkins or Kubernetes servers
handling ProwJobs, the plank sync loop can take hours or longer as each
ProwJob is synced serially and client timeouts in each sync add up. Now,
the sync is batched with a limit of 20 concurrent routines at once.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

Fixes https://github.com/kubernetes/test-infra/issues/3742

/area prow
/cc @spxtr

### CODE REVIEW

**Comment by stevekuznetsov** [comment]
@spxtr should the GCS bucket have more artifacts? I'm seeing this in the build log:

```
W0727 20:14:04.738]         Testing //prow/plank:go_default_test, 269 s
W0727 20:14:35.632] TIMEOUT: //prow/plank:go_default_test (see /root/.cache/bazel/_bazel_root/27a95fe2fc68aa1195ddf1a4bbaaea62/execroot/test-infra/bazel-out/local-fastbuild/testlogs/prow/plank/go_default_test/test.log)
```

But the test log is nowhere to be found

**Comment by stevekuznetsov** [comment]
Running it locally I see...

```
$ cat ~/.cache/bazel/_bazel_skuznets/fe650a89676a24b8d48936a0c5c19473/execroot/__main__/bazel-out/local-fastbuild/testlogs/prow/plank/go_default_test/test.log
exec ${PAGER:-/usr/bin/less} "$0" || exit 1
-----------------------------------------------------------------------------
time="2017-07-27T20:32:07Z" level=info msg="Not starting another instance of same, already 1 running." job=same 
```

How to look at `go test` output? I'm assuming that is what is hanging?

File: prow/plank/controller.go [hunk 1]
```diff
@@ -117,19 +122,25 @@ func (c *Controller) Sync() error {
...
-				syncErrs = append(syncErrs, err)
-			}
-		} else if pj.Spec.Agent == kube.JenkinsAgent {
-			if err := c.syncJenkinsJob(pj, pending); err != nil {
-				syncErrs = append(syncErrs, err)
-			}
-		} else {
-			syncErrs = append(syncErrs, fmt.Errorf("job %s has unsupported agent %s", pj.Metadata.Name, pj.Spec.Agent))
-		}
+		pjCh <- pj
```
Comment (spxtr) [inline prow/plank/controller.go]: This is blocking.

**Comment by spxtr** [comment]
The `pending` variable is a raw map that will be updated by the sync calls. It needs to be protected by a lock.

Other than those two issues, this looks good.

File: prow/plank/controller.go [hunk 2]
```diff
@@ -142,6 +153,24 @@ func (c *Controller) Sync() error {
...
+		if pj.Spec.Agent == kube.KubernetesAgent {
+			if err := c.syncKubernetesJob(pj, pm, pending); err != nil {
+				syncErrors <- err
+			}
+		} else if pj.Spec.Agent == kube.JenkinsAgent {
+			if err := c.syncJenkinsJob(pj, pending); err != nil {
+				syncErrors <- err
+			}
+		} else {
+			syncErrors <- fmt.Errorf("job %s has unsupported agent %s", pj.Metadata.Name, pj.Spec.Agent)
```
Comment (kargakis) [inline prow/plank/controller.go]: This should be relaxed... In the future each agent will need to be handled by a separate controller and having the main controller (plank) ignore unknown agents will actually make it possible for the system to be extendable.

Comment (stevekuznetsov) [inline prow/plank/controller.go]: Seems unrelated

Comment (spxtr) [inline prow/plank/controller.go]: Agree with both.

Comment (kargakis) [inline prow/plank/controller.go]: I mean it should not be regarded as an error here. It's definitely not unrelated if we change the controller to use informers and resync prowJobs on errors.

**Comment by stevekuznetsov** [comment]
Something other than the `pending` map is racing here as well. Digging into it.

**Comment by stevekuznetsov** [comment]
OK, addressed the races. @spxtr does Bazel run these tests with `-race` ?

File: prow/plank/controller.go [hunk 3]
```diff
@@ -112,24 +158,31 @@ func (c *Controller) Sync() error {
...
-				syncErrs = append(syncErrs, err)
-			}
-		} else {
-			syncErrs = append(syncErrs, fmt.Errorf("job %s has unsupported agent %s", pj.Metadata.Name, pj.Spec.Agent))
-		}
+		pjCh <- pj
 	}
+	close(pjCh)
+
+	errCh := make(chan error, len(pjs))
```
Comment (stevekuznetsov) [inline prow/plank/controller.go]: When someone is refactoring, if they push more than one `error` per `ProwJob` into this they can deadlock the controller. Do we want a stronger guarantee that this won't happen?

File: prow/plank/controller.go [hunk 4]
```diff
@@ -77,22 +82,63 @@ type Controller struct {
...
+
+	t.pendingJobs[job] = t.pendingJobs[job] + 1
+}
+
+// SetPending sets the currently pending set of jobs.
+// This is NOT thread safe and should only be used for
+// initializing the controller during testing.
+func (c *Controller) SetPending(pendingJobs map[string]int) {
+	if pendingJobs == nil {
+		c.pending.pendingJobs = make(map[string]int)
```
Comment (kargakis) [inline prow/plank/controller.go]: Shouldn't you set pendingJobs here, too?

File: prow/plank/controller.go [hunk 5]
```diff
@@ -77,22 +82,63 @@ type Controller struct {
...
+	defer t.lock.RUnlock()
+	return t.pendingJobs[key]
+}
+
+// Add increments the amount of pending ProwJobs
+// for the given job identifier
+func (t *pendingJobsTracker) Increment(job string) {
+	t.lock.RLock()
+	defer t.lock.RUnlock()
+
```
Comment (kargakis) [inline prow/plank/controller.go]: Check that t.pendingJobs is nil

File: prow/plank/controller.go [hunk 6]
```diff
@@ -77,22 +82,63 @@ type Controller struct {
...
+func (t *pendingJobsTracker) Get(key string) int {
+	t.lock.RLock()
+	defer t.lock.RUnlock()
+	return t.pendingJobs[key]
+}
+
+// Add increments the amount of pending ProwJobs
+// for the given job identifier
+func (t *pendingJobsTracker) Increment(job string) {
+	t.lock.RLock()
```
Comment (kargakis) [inline prow/plank/controller.go]: You probably want to use t.lock.Lock() for writes.

File: prow/plank/controller.go [hunk 7]
```diff
@@ -77,22 +82,63 @@ type Controller struct {
...
+	pending pendingJobsTracker
+
 	reports []kube.ProwJob
 }
 
+// pendingJobsTracker guards a mapping of pending
+// jobs so we can share it between goroutines
+type pendingJobsTracker struct {
+	pendingJobs map[string]int
+	lock        sync.RWMutex
```
Comment (kargakis) [inline prow/plank/controller.go]: I like to embed this in simple structs.

File: prow/plank/controller.go [hunk 8]
```diff
@@ -112,24 +158,31 @@ func (c *Controller) Sync() error {
...
 		pm[pod.Metadata.Name] = pod
 	}
-	pending := pendingJobs(pjs)
+	c.updatePendingJobs(pjs)
 	var syncErrs []error
 	if err := c.terminateDupes(pjs); err != nil {
 		syncErrs = append(syncErrs, err)
 	}
+
+	pjCh := make(chan kube.ProwJob, len(pjs))
```
Comment (kargakis) [inline prow/plank/controller.go]: Not clear why we need this channel. Isn't it simpler to run the second loop you run directly?

File: prow/plank/controller.go (hunk 5 above)
Comment (stevekuznetsov) [inline prow/plank/controller.go]: Why not require it to be well formed from `Initialize`?

File: prow/plank/controller.go (hunk 4 above)
Comment (stevekuznetsov) [inline prow/plank/controller.go]: Not sure what you mean

File: prow/plank/controller.go (hunk 8 above)
Comment (stevekuznetsov) [inline prow/plank/controller.go]: How are the 20 workers going to be fed the jobs to work on?

**Comment by stevekuznetsov** [comment]
@kargakis updated

File: prow/plank/controller.go (hunk 8 above)
Comment (spxtr) [inline prow/plank/controller.go]: I think this is a neat solution. You can use a semaphore and just make a goroutine per job, but I like this more.

**Comment by spxtr** [comment]
> OK, addressed the races. @spxtr does Bazel run these tests with `-race` ?

Not yet, I'm working on that. In the meantime please test locally with `-race`.

File: prow/plank/controller.go [hunk 9]
```diff
@@ -77,23 +82,56 @@ type Controller struct {
...
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	c.pendingJobs[job] = c.pendingJobs[job] + 1
+}
+
+// SetPending sets the currently pending set of jobs.
+// This is NOT thread safe and should only be used for
+// initializing the controller during testing.
+func (c *Controller) SetPending(pendingJobs map[string]int) {
```
Comment (spxtr) [inline prow/plank/controller.go]: Can you make this private?

**Comment by stevekuznetsov** [comment]
@spxtr with `-race`:

```
$ bazel test //prow/plank:go_default_test --test_output=all --features=race
..........
INFO: Found 1 test target...
INFO: From Testing //prow/plank:go_default_test:
==================== Test output for //prow/plank:go_default_test:
time="2017-07-28T03:24:28Z" level=info msg="Not starting another instance of same, already 1 running." job=same 
PASS
================================================================================
Target //prow/plank:go_default_test up-to-date:
  bazel-bin/prow/plank/go_default_test
INFO: Elapsed time: 9.630s, Critical Path: 3.67s
//prow/plank:go_default_test                                             PASSED in 1.1s

Executed 1 out of 1 test: 1 test passes.
```

File: prow/plank/controller.go (hunk 9 above)
Comment (stevekuznetsov) [inline prow/plank/controller.go]: Done

File: prow/plank/controller.go (hunk 4 above)
Comment (kargakis) [inline prow/plank/controller.go]: Late night comment. Looks good

File: prow/plank/controller.go (hunk 8 above)
Comment (kargakis) [inline prow/plank/controller.go]: Yeah, this is better, disregard

**Comment by spxtr** [comment]
Looks good. Please run `./bump.sh plank` and add the result to the PR, then I'll merge :)

/lgtm

**Comment by stevekuznetsov** [comment]
@spxtr `plank` version bumped

**Comment by spxtr** [comment]
OK!
